mlflow/utils/requirements_utils.py

```diff
@@ -541,7 +541,18 @@ def _infer_requirements(model_uri, flavor, raise_on_error=False, extra_env_vars=
             unrecognized_packages,
         )
 
-    return sorted(map(_get_pinned_requirement, packages))
+    packages = set(map(_get_pinned_requirement, packages))
+
+    # Handle pandas incompatibility issue with numpy 2.x https://github.com/pandas-dev/pandas/issues/55519
+    # pandas == 2.2.*: compatible with numpy >= 2
+    # pandas >= 2.1.2: incompatible with numpy >= 2, but it pins numpy < 2
+    # pandas < 2.1.2: incompatible with numpy >= 2 and doesn't pin numpy, so we need to pin numpy
+    if any(
+        package.startswith("pandas") and Version(package.split("==")[1]) < Version("2.1.2")
```

Comment:
```suggestion
        package == "pandas" and Version(package.split("==")[1]) < Version("2.1.2")
```

is there a reason that this doesn't work?

Reply:
Here package is `pandas==xxx` because we have applied _get_pinned_requirement on it, otherwise we need to check before the mapping

Reply:
Got it. Let's fix this. We should pin numpy when packages contains `pandasflow`.